File: src/main/java/dev/obscuria/tooltips/client/tooltip/filter/NoneOfFilter.java
package dev.obscuria.tooltips.client.tooltip.filter;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import net.minecraft.item.ItemStack;

import java.util.List;

@Desugar
public record NoneOfFilter(List<ItemFilter> terms) implements ItemFilter {

    public static NoneOfFilter fromJson(JsonObject json) {
        return new NoneOfFilter(ItemFilter.parseTerms(json));
    }

    @Override
    public boolean test(ItemStack stack) {
        return terms.stream().noneMatch(term -> term.test(stack));
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/filter/NoneOfFilter.kt
package dev.obscuria.tooltips.client.filter

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.world.item.ItemStack

class NoneOfFilter(
    val terms: List<IItemFilter>
) : IItemFilter {

    override fun codec(): Codec<NoneOfFilter> = CODEC

    override fun test(stack: ItemStack): Boolean {
        return terms.none { it.test(stack) }
    }

    companion object {

        val CODEC: Codec<NoneOfFilter> = RecordCodecBuilder.create { codec ->
            codec.group(
                IItemFilter.CODEC.listOf()
                    .fieldOf("terms")
                    .forGetter(NoneOfFilter::terms)
            ).apply(codec, ::NoneOfFilter)
        }
    }
}